utils/src/main/java/software/amazon/awssdk/utils/ReflectionMethodInvoker.java

```diff
@@ -89,21 +91,32 @@ public R invoke(T obj, Object... args) throws NoSuchMethodException {
         }
     }
 
-    private synchronized Method getTargetMethod() throws NoSuchMethodException {
-        if (cachedException != null) {
-            throw cachedException;
-        }
+    /**
+     * Initializes the method invoker by finding and caching the target method from the target class that will be used
+     * for subsequent invoke operations. If the invoker is already initialized this call will do nothing.
+     * @throws NoSuchMethodException if the JVM could not find a method matching the signature specified in the
+     * initialization of this proxy.
+     */
+    public void initialize() throws NoSuchMethodException {
+        getTargetMethod();
+    }
```

Comment:
Is this backwards-compatible? We usually remain backwards-compatible with SdkProtectedApi classes like these.

Reply:
Mostly. Invoke() will initialize the object even if initialize() is not explicitly called, and so will behave as the older version did. What has changed, however, is that invoke() used to cache a thrown NoSuchMethodException and it no longer does that which will result in a small efficiency drop in cases where it is repeatedly called and the method does not exist. Functionally it behaves the same, just with additional features/controls.